flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/ExactlyOnceKafkaWriter.java

```diff
@@ -136,13 +158,10 @@ public void initialize() {
     }
 
     private FlinkKafkaInternalProducer<byte[], byte[]> startTransaction(long checkpointId) {
+        namingContext.setNextCheckpointId(checkpointId);
         FlinkKafkaInternalProducer<byte[], byte[]> producer =
-                producerPool.getTransactionalProducer(
-                        TransactionalIdFactory.buildTransactionalId(
-                                transactionalIdPrefix,
-                                kafkaSinkContext.getParallelInstanceId(),
-                                checkpointId),
-                        checkpointId);
+                transactionNamingStrategy.getTransactionalProducer(namingContext);
+        namingContext.setLastCheckpointId(checkpointId);
```

Comment:
This looks strange. Why do we expose `setNextCheckpointId` and `setLastCheckpointId` if both methods are always used together? Is the interleaving of the producer creation necessary?

Reply:
Not sure how to improve it. There could be an arbitrary number of aborted transactions between the two invocations of the call-site. For me, these are two disjunct states that are coupled in the best case but not in the general case. Remember that we need to `setLastCheckpointId` once already during recovery.